lib/spack/spack/config.py

```diff
@@ -230,32 +234,48 @@ def get_section(self, section):
             if self._raw_data is None:
                 return None
 
+            section_data = copy.deepcopy(self._raw_data)
             for key in self.yaml_path:
-                if self._raw_data is None:
+                if section_data is None:
                     return None
 
-                # support tuples as "or" in the yaml path
-                if isinstance(key, (list, tuple)):
-                    key = first_existing(self._raw_data, key)
-
-                self._raw_data = self._raw_data[key]
+                section_data = section_data[key]
 
-            for section_key, data in self._raw_data.items():
+            for section_key, data in section_data.items():
                 self.sections[section_key] = {section_key: data}
-
         return self.sections.get(section, None)
 
     def write_section(self, section):
-        validate(self.sections, self.schema)
+        data_to_write = copy.deepcopy(self._raw_data)
```

Comment:
This logic is for preserving comments in `spack.yaml` IMO it could use a test like `test_config_update_can_handle_comments` in `test/cmd/config`.

Also, I'm not clear why but in the analogous logic for `ConfigScope`, `deepcopy` is only used on `write`, so I'd have to investigate why that is needed here. Do you have a notion of what goes wrong when that is not used here?

Reply:
That is tested for in the tests that update an environment config. Deepcopy is used to ensure that the modifications we make below don't modify the _raw_data object when we chop off the keys that aren't part of the actual config data (self.yaml_path). Since normal `ConfigScope` objects don't have to do that chopping, they don't need to use `copy.deepcopy` on read.